opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,249 @@
+
+    @property
+    def connector(self):
+        """Define the initiated matrix connector."""
+        return self.opsdroid.get_connector("matrix")
+
+    @property
+    def room_id(self):
+        """Define ID of the room to store data in."""
+        return (
+            self.room
+            if self.room.startswith("!")
+            else self.connector.room_ids[self.room]
+        )
+
+    async def migrate_database(self):
+        """Migrate existing 'opsdroid.database' state event to 'dev.opsdroid.database' event."""
+
+        data = await self.connector.connection.room_get_state(room_id=self.room_id)
+        if isinstance(data, RoomGetStateError):
+            _LOGGER.error(
+                f"Error migrating from opsdroid.database to {self._event_type} in room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+
+        if not any(event["type"] == self._event_type for event in data.events):
+            for event in data.events:
+                if event["type"] == "opsdroid.database":
+                    await self.connector.connection.room_put_state(
+                        room_id=self.room_id,
+                        event_type=self._event_type,
+                        content=event["content"],
+                        state_key=event["state_key"],
+                    )
+                    await self.connector.connection.room_redact(
+                        room_id=self.room_id, event_id=event["event_id"],
+                    )
+
+        self.should_migrate = False
+
+    async def connect(self):
+        """Connect to the database."""
+
+        _LOGGER.info("Matrix Database connector initialised.")
+
+    async def put(self, key, value):
+        """Insert or replace a value into the database for a given key."""
+
+        if self.should_migrate:
+            await self.migrate_database()
+
+        # If the single state key flag is set then use that else use state key.
+        state_key = (
+            "" if self._single_state_key is True else self._single_state_key or key
+        )
+
+        _LOGGER.debug(f"Putting {key} into matrix room {self.room_id}.")
+
+        ori_data = await self.connector.connection.room_get_state_event(
+            room_id=self.room_id, event_type=self._event_type, state_key=state_key,
+        )
+        if isinstance(ori_data, RoomGetStateEventError):
+            _LOGGER.error(
+                f"Error putting key into matrix room {self.room_id}: {ori_data.message}({ori_data.status_code})"
+            )
+            return
+        ori_data = ori_data.content
+
+        if self._single_state_key:
+            value = {key: value}
+
+        elif not isinstance(value, dict):
+            raise ValueError(
+                "When the matrix database is configured with single_state_key=False, "
+                "the value passed must be a dict."
+            )
+
+        sub_key = list(value.keys())[0]
```

Comment:
this doesn't make any sense to me, if `single_state_key` is `False` then there can be an arbitrary number of keys in `value`.